### DISCUSSION TITLE
Useful to have an API to parse text files with rules and queries

### DISCUSSION THREAD

**Discussion by pvillads** [microsoft/Guan#46] [Ideas]
Today, the user has to provide a list of strings as shown:
            var logicRules2 = new List<string>
            {
                "person(adam)",
                "person(betty)",
                "person(carl)",
                "person(dan)",
                "test(?x) :- not(person(?x)), WriteInfo('{0} is not a person', ?x)",
                "test(?x) :- WriteInfo('{0} is a person', ?x)"
             };

and then parse the rules in that list with the 

Module module2 = Module.Parse("persontests", logicRules2, null);

API. In addition to this API I think it would be useful to have an API where the parser did this work for me, possibly aided by a '.' at the end of each fact / rule. The API would then be:

var logicRules3 = 
@"person(adam).
    person(betty).
    person(carl).
    person(dan).

    test(?x) :- not(person(?x)), WriteInfo('{0} is not a person', ?x).
    test(?x) :- WriteInfo('{0} is a person', ?x).";

var module = Module.Parse("PersonTests", logicRules3, null);

If this is not the case, consumers of Guan need to come up with all sorts of heuristics that deal with delimiting rules, typically based on  splitting on newline, which imposes unacceptable compromises to how the user can format complex rules. And, that sort of parsing should really not have to be done by consumers, but by the guan parser, IMO.

BR PV

**Comment by GitTorre** [comment]
Hi,

Thanks for the feedback.

Guan expects a List of string. The consuming code needs to produce that list. See FabricHealer's implementation for a sample of how to read rules from a file into a list: 

https://github.com/microsoft/service-fabric-healer/blob/1e8a4ebb9818254938bab78f8bb92dc111a86a83/FabricHealer/FabricHealerManager.cs#L1923

which calls: 

https://github.com/microsoft/service-fabric-healer/blob/1e8a4ebb9818254938bab78f8bb92dc111a86a83/FabricHealer/FabricHealerManager.cs#L1995

As you can see, the application above (FabricHealer) employs trailing "." to delimit individual rules for the specific purpose of feeding the rules from a file into a ```List<string>```. FabricHealer also employs comments for rules (with "##" lines) that are removed by the custom pre-parser.

**Reply by TaviTruman** [reply]
@GitTorre Thanks for sharing this bit of knowledge and the reference to the implementation; I am using Guan to store logic to process OWL 2 graphs. I store Guan logic strings in our Trinity Graph Engine services. Bravo - excellent library. We need extended expressiveness in processing ISO Common Logic (FOL, SOL, etc.) productions.
